KAFKA-14225: Fix deadlock caused by lazy val exemptSensor

hshi2022 opened the pull request:
There is a chance to cause deadlock when multiple threads access ClientRequestQuotaManager. In the version Scala 2.12, the lazy val initialization is under the object lock. The deadlock could happen in the following condition:

In thread a, when ClientRequestQuotaManager.exemptSensor is being initialized, it has acquired the object lock and enters the the actual initialization block. The initialization of 'exemptSensor' requires another lock private val lock = new ReentrantReadWriteLock() and it is waiting for this lock.

In thread b, at the same time, ClientQuotaManager.updateQuota() is called and it has already acquired ReentrantReadWriteLock lock by calling lock.writeLock().lock(). And then it executes info(). If this is the first time accessing Logging.logger, which is also a lazy val, it need to wait for the object lock.

The deadlock happens.

Since the lazy val initialization is under the object lock, we should avoid using lazy val if the initialization function holds another lock to prevent holding two locks at the same time which is prone for deadlock. Change to create exemptSensor during ClientRequestQuotaManager initialization with an expiration time of Long.MaxValue to prevent expiration if request quota is not enabled at that time.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

gitlw commented:
rerun tests please

hshi2022 committed:
There is a chance to cause deadlock when multiple threads access ClientRequestQuotaManager. In the version Scala 2.12, the lazy val initialization is under the object lock. The deadlock could happen in the following condition:

In thread a, when ClientRequestQuotaManager.exemptSensor is being initialized, it has acquired the object lock and enters the the actual initialization block. The initialization of 'exemptSensor' requires another lock private val lock = new ReentrantReadWriteLock() and it is waiting for this lock.

In thread b, at the same time, ClientQuotaManager.updateQuota() is called and it has already acquired ReentrantReadWriteLock lock by calling lock.writeLock().lock(). And then it executes info(). If this is the first time accessing Logging.logger, which is also a lazy val, it need to wait for the object lock.

The deadlock happens.

Since the lazy val initialization is under the object lock, we should avoid using lazy val if the initialization function holds another lock to prevent holding two locks at the same time which is prone for deadlock. Change to create exemptSensor during ClientRequestQuotaManager initialization with an expiration time of Long.MaxValue to prevent expiration if request quota is not enabled at that time.

hshi2022 commented:
@dajac @rajinisivaram could you please review this PR? Thanks!

hachikuji commented:
@hshi2022 Thanks for the patch. Did you catch this deadlock in practice or did you use some tool to find it?

hshi2022 commented:
> @hshi2022 Thanks for the patch. Did you catch this deadlock in practice or did you use some tool to find it?

We caught in practice.

hachikuji commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -433,10 +433,10 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
   }
 
-  protected def getOrCreateSensor(sensorName: String, metricName: MetricName): Sensor = {
+  protected def getOrCreateSensor(sensorName: String, expirationTime: Long, metricName: MetricName): Sensor = {
```
nit: `expirationTimeMs`? It's conventional to include the unit type (even though it's almost always milliseconds). 

By the way, this method feels a bit clumsy. It assumes that the provided metric name describes a "rate" even though there is nothing in the name of the method that suggests it. It would be clearer to leave the metric initialization to the caller. For example:
```scala
protected def getOrCreateSensor(
  sensorName: String,
  expirationTime: Long,
  metricRegistrationCallback: Sensor => Unit
): Unit = {
    sensorAccessor.getOrCreate(
      sensorName,
      expirationTime,
      metricRegistrationCallback
    )
}
```
Then the (only) caller can do the following:
```scala
val exemptSensor: Sensor = getOrCreateSensor(
  sensorName = ClientRequestQuotaManager.ExemptSensorName, 
  expirationTimeMs = Long.MaxValue,
  metricRegistrationCallback = sensor => sensor.add(exemptMetricName, new Rate)
)
```

hshi2022 replied:
update as suggested

hachikuji commented on core/src/main/scala/kafka/server/ClientRequestQuotaManager.scala:
```diff
@@ -30,6 +30,9 @@ import scala.jdk.CollectionConverters._
   val NanosToPercentagePerSecond = 100.0 / TimeUnit.SECONDS.toNanos(1)
+  // Since exemptSensor is for all clients and has a constant name, we do not expire exemptSensor and only
+  // create once.
+  val DefaultInactiveExemptSensorExpirationTimeSeconds = Long.MaxValue
```
nit: perhaps overkill to have a constant. I'd suggest
```scala
  // Since exemptSensor is for all clients and has a constant name, we do not expire it
  val exemptSensor: Sensor = getOrCreateSensor(ClientRequestQuotaManager.ExemptSensorName,
    Long.MaxValue, exemptMetricName)
```

hshi2022 replied:
I prefer creating exemptSensor at initialization because getOrCreateSensor acquires lock and if creating exemptSensor every time it is used may have negative performance impact. And since it is just one metric, it should be good to create it during initialization and have it never expired (not a big issue for space usage).

hachikuji replied:
To clarify, I'm just suggesting we get rid of `DefaultInactiveExemptSensorExpirationTimeSeconds` and use `Long.MaxValue` directly. We already have the comment which emphasizes that the sensor does not get expired.

```scala
// Since exemptSensor is for all clients and has a constant name, we do not expire it
val exemptSensor: Sensor = getOrCreateSensor(
  sensorName = ClientRequestQuotaManager.ExemptSensorName, 
  expirationTimeMs = Long.MaxValue,
  metricRegistrationCallback = sensor => sensor.add(exemptMetricName, new Rate)
)
```

hshi2022 replied:
I prefer using DefaultInactiveExemptSensorExpirationTimeSeconds because the value could be updated if needed. And it aligns with the code pattern, e.g., we also have QuotaRequestPercentDefault which has Int.MaxValue

hachikuji reviewed (commented):
Thanks, the fix seems reasonable. I left a couple minor comments.

hshi2022 committed:
address comments

hshi2022 commented:
@hachikuji Thanks for review. I have addressed your comments. And I am wondering whether you know how to rerun the checks/tests without submitting any changes. This PR currently fails a test, which I do not think is related to the PR's change.)

hachikuji commented:
@hshi2022 Yeah, we are often fighting some flaky tests. It should not block merging if they are unrelated. Can you take a look at the response above https://github.com/apache/kafka/pull/12634#discussion_r982935129? Thanks!

hshi2022 commented:
> @hshi2022 Yeah, we are often fighting some flaky tests. It should not block merging if they are unrelated. Can you take a look at the response above [#12634 (comment)](https://github.com/apache/kafka/pull/12634#discussion_r982935129)? Thanks!

hshi2022 commented:
> @hshi2022 Yeah, we are often fighting some flaky tests. It should not block merging if they are unrelated. Can you take a look at the response above [#12634 (comment)](https://github.com/apache/kafka/pull/12634#discussion_r982935129)? Thanks!

Replied under that comment.

hshi2022 commented:
@hachikuji Do you have any more questions/comments regarding this PR?

hachikuji reviewed (approved):
Thanks, LGTM

hachikuji commented:
Kicked off one more build since it's been a little while. I will merge assuming nothing wrong with the build. Thanks for the patch!
